src/osd/OSD.h

```diff
@@ -1094,6 +1021,221 @@ class OSDService {
+    it to the same_interval_since epoch] so that we don't wait for epochs that
+    don't affect the given PG.)
+
+  - we maintain two separate wait lists, *waiting* and *waiting_peering*. The
+    OpQueueItem has an is_peering() bool to determine which we use.  Waiting
+    peering events are queued up by epoch required.
+
+  - when we wake a PG slot (e.g., we finished split, or got a newer osdmap, or
+    materialized the PG), we wake *all* waiting items.  (This could be optimized,
+    probably, but we don't bother.)  We always requeue peering items ahead of
+    client ops.
+
+  - some peering events are marked !peering_requires_pg (PGQuery).  if we do
+    not have a PG these are processed immediately (under the shard lock).
+
+  - we do not have a PG present, we check if the slot maps to the current host.
+    if so, we either queue the item and wait for the PG to materialize, or
+    (if the event is a pg creating event like PGNotify), we materialize the PG.
+
+  - when we advance the osdmap on the OSDShard, we scan pg slots and
+    discard any slots with no pg (and not waiting_for_split) that no
+    longer map to the current host.
+
+  Some notes:
+
+  - There is theoretical race between query (which can proceed if the pg doesn't
+    exist) and split (which may materialize a PG in a different shard):
+      - osd has epoch E
+      - shard 0 processes notify on P from epoch E-1
+      - shard 0 identifies P splits to P+C in epoch E
+      - shard 1 receives query for P (epoch E), returns DNE
+      - shard 1 installs C in shard 0 with waiting_for_split
+
+    This can't really be fixed without ordering queries over all shards.  In
+    practice, it is very unlikely to occur, since only the primary sends a
+    notify (or other creating event) and only the primary who sends a query.
+    Even if it does happen, the instantiated P is empty, so reporting DNE vs
+    empty C is minimal harm.
+
+  */
+
+struct OSDShardPGSlot {
+  PGRef pg;                      ///< pg reference
+  deque<OpQueueItem> to_process; ///< order items for this slot
+  int num_running = 0;          ///< _process threads doing pg lookup/lock
+
+  deque<OpQueueItem> waiting;   ///< waiting for pg (or map + pg)
+
+  /// waiting for map (peering evt)
+  map<epoch_t,deque<OpQueueItem>> waiting_peering;
+
+  /// incremented by wake_pg_waiters; indicates racing _process threads
+  /// should bail out (their op has been requeued)
+  uint64_t requeue_seq = 0;
+
+  /// waiting for split child to materialize
+  bool waiting_for_split = false;
+
+  epoch_t epoch = 0;
+  boost::intrusive::set_member_hook<> pg_epoch_item;
+};
+
+struct OSDShard {
+  const unsigned shard_id;
+  CephContext *cct;
+  OSD *osd;
+
+  string shard_name;
+
+  string sdata_lock_name;
+  Mutex sdata_lock;
+  Cond sdata_cond;
+
+  string sdata_op_ordering_lock_name;
+  Mutex sdata_op_ordering_lock;   ///< protects all members below
+
+  string osdmap_lock_name;
+  Mutex osdmap_lock;
```

Comment:
I'm really uncertain just how this is supposed to be protecting the shard's ```osdmap``` member. There are still several unguarded accesses to it, not all of which are in ```OSDShard::consume_map()```. (At minimum, initialization and shutdown, but I think some others as well.)

Reply:
consume_map, init, and shutdown aren't racing with updates.  The problem was just with _process (and dequeue_peering_evt in particular) racing with consume_map update the ref at the same time.

Reply:
```identify_splits()```, ```prime_splits()```, ```_process()``` all take unguarded references to it. Maaaaybe the only conflict to worry about is ```consume_map()``` and ```dequeue_peering_evt()```, but it's definitely not an ```osdmap_lock``` in that case!

Reply:
Ah, it's because those are all protected by shard_lock (or are init/teardown).  dequeue_peering_evt is the only one that uses it without shard_lock, and consume_map has the one site that updates it.. these are the only two accesses that need protection from each other.  (Adding locking for init/shutdown for good hygiene.)